### PR TITLE
metrics: shows formatted multiline metrics

### LINKED ISSUE #1716: Prettify TSV metrics output
When displaying TSV metrics output the printout is not readable:

```
	../../20181223-TrainSetZero/SanityCheck/Bravo_on_TrainSetZero.metrics: value_mse	deviation_mse	data_set
0.421601	0.173461	train
0.67528	0.289545	testing
0.671502	0.297848	validation
```

I think it would be much eaier to read if a newline were added, and the remaining text were formatted as though it had been passed via `column -t`:

```
	../../20181223-TrainSetZero/SanityCheck/Bravo_on_TrainSetZero.metrics:
value_mse  deviation_mse  data_set
0.421601   0.173461       train
0.67528    0.289545       testing
0.671502   0.297848       validation
```

### PR DESCRIPTION
**Pull request by mroutis**
Close #1716

----

How things should look like:
```
❯ dvc metrics show
	metrics.tsv:
		value_mse   deviation_mse   data_set
		0.421601    0.173461        train
		0.67528     0.289545        testing
		0.671502    0.297848        validation
	metrics.csv:
		value_mse   deviation_mse   data_set
		0.421601    0.173461        train
		0.67528     0.289545        testing
		0.671502    0.297848        validation
	metrics.txt:
		ROC_AUC:   0.64
		KS:        78.9999999996
		F_SCORE:   77
	metrics.json:
		{
		   "deviation_mse": [
		      "0.173461",
		      "0.289545",
		      "0.297848"
		   ],
		   "data_set": [
		      "train",
		      "testing",
		      "validation"
		   ],
		   "value_mse": [
		      "0.421601",
		      "0.67528",
		      "0.671502"
		   ]
		}
```

### CODE REVIEW

File: dvc/command/metrics.py [hunk 1]
```diff
@@ -12,7 +12,32 @@ def _show(self, metrics):
                 logger.info("{}:".format(branch))
 
             for fname, metric in val.items():
-                logger.info("\t{}: {}".format(fname, metric))
+                self._output_formated_metric(fname, metric)
+
+    def _output_formated_metric(self, fname, metric):
```
Comment (shcheklein) [inline dvc/command/metrics.py]: minor: formatted

File: dvc/repo/metrics/show.py [hunk 2]
```diff
@@ -136,7 +136,7 @@ def _read_metrics(self, metrics, branch):
         if not metric:
             continue
 
-        res[out.rel_path] = metric
+        res[out.rel_path] = (metric, typ)
```
Comment (mroutis) [inline dvc/repo/metrics/show.py]: @efiop , is there a better way to achieve this?

File: dvc/command/metrics.py [hunk 3]
```diff
@@ -12,7 +12,32 @@ def _show(self, metrics):
...
+            typ
+        )
+
+        if not delimiter or "\n" not in content:
+            logger.info("\t{}: {}".format(fname, content))
+            return
+
+        logger.info("\t{}:".format(fname))
+
+        rows = [line.split(delimiter) for line in content.split("\n")]
```
Comment (shcheklein) [inline dvc/command/metrics.py]: this won't work for CSV as far as I remember. CSV entry may contain delimiters, newlines, etc. The only proper way to split CSV is to use CSV reader.

Comment (mroutis) [reply dvc/command/metrics.py]: @shcheklein, csv wasn't working that well, I tried `csv.reader([content], delimiter=builtin_str(delimiter))` and was asking something about universal newline.

I assumed that a simple case like some metrics separated by a comma wouldn't break this so often.

File: dvc/command/metrics.py [hunk 4]
```diff
@@ -12,7 +12,32 @@ def _show(self, metrics):
...
+                self._output_formated_metric(fname, metric)
+
+    def _output_formated_metric(self, fname, metric):
+        content, typ = metric
+
+        delimiter = {"csv": ",", "hcsv": ",", "tsv": "\t", "htsv": "\t"}.get(
+            typ
+        )
+
+        if not delimiter or "\n" not in content:
```
Comment (shcheklein) [inline dvc/command/metrics.py]: in what cases do we need this `if`? (I know that sometimes content is actually should be a list, are there actual content (single line, or without delimiters) that could not be processed with code below (may be fixed a bit))?

File: dvc/command/metrics.py (hunk 3 above)
Comment (shcheklein) [reply dvc/command/metrics.py]: I'm not sure I understood why wasn't it working well, could you elaborate a little bit, please?

still, I think if support CSV we must parse it properly, otherwise you can unexpected results and bugs.

For example, will zip(*rows) work if rows have different size (and this will happen if there entries that contain , or \n).

File: tests/test_metrics.py [hunk 5]
```diff
@@ -167,6 +167,53 @@ def test_xpath_all_with_header(self):
...
+
+        self.dvc.metrics.modify("metrics.csv", typ="csv")
+        self.dvc.metrics.modify("metrics.tsv", typ="tsv")
+
+        with MockLoggerHandlers(logger.logger):
+            reset_logger_standard_output()
+
+            ret = main(["metrics", "show"])
+
+            expected_output = (
```
Comment (shcheklein) [inline tests/test_metrics.py]: love the test! please, let's add a simple test for an edge case - no delimiters or no newlines

File: dvc/command/metrics.py (hunk 4 above)
Comment (mroutis) [reply dvc/command/metrics.py]: Because we don't want to process `json`s nor `raw` files.

Also, if the output is a one-liner, it could fit on the same line to be consistent with `raw`:
```
metrics.csv: 0.36,0.88
```

File: dvc/repo/metrics/show.py (hunk 2 above)
Comment (mroutis) [reply dvc/repo/metrics/show.py]: I could use `https://docs.python.org/2/library/csv.html#csv.Sniffer` to detect if the file is separated by comas, semicolons, or tabs.

So I don't need the type at all :thinking:

File: dvc/command/metrics.py (hunk 3 above)
Comment (mroutis) [reply dvc/command/metrics.py]:
```python
reader = csv.reader([content], delimiter=delimiter)

TypeError: '"delimiter" must be string, not unicode'
```

It doesn't support unicode as a delimiter, so we need to use `builtin_str`

```python
content = u'value_mse\tdeviation_mse\tdata_set\n0.421601\t0.173461\ttrain\n0.67528\t0.289545\ttesting\n0.671502\t0.297848\tvalidation'

reader = csv.reader([content], delimiter=builtin_str(delimiter))
rows = [row for row in reader]

*** Error: new-line character seen in unquoted field - do you need to open the file in universal-newline mode?
```

It expects that the content is already _split_ by new lines.

But indeed, it is more reliable because of quoting :+1: .

File: tests/test_metrics.py (hunk 5 above)
Comment (mroutis) [reply tests/test_metrics.py]: I'm missing this one, will work on it tomorrow :)

File: dvc/repo/metrics/show.py (hunk 2 above)
Comment (mroutis) [reply dvc/repo/metrics/show.py]: Need to test for "raw" type with commas and "json" with comas (maybe adding some more cases).

Dang, maybe I should've created different commits instead of amending :thinking:

File: dvc/command/metrics.py (hunk 3 above)
Comment (shcheklein) [reply dvc/command/metrics.py]: if you want to use CSV reader to read TSV most likely you need to setup it properly - disable, escaping rules, quoting of fields, etc. I would go with simple split for TSV and a reader for CSV.

File: dvc/command/metrics.py [hunk 6]
```diff
@@ -12,7 +12,46 @@ def _show(self, metrics):
...
+        try:
+            json.loads(content)
+            return True
+        except ValueError:
+            return False
+
+    def _output_formatted_metric(self, fname, metric):
+        import csv
+
+        if "\n" not in metric:
```
Comment (shcheklein) [inline dvc/command/metrics.py]: I think that metric could be an array when `xpath` option is provided and "\n" could be potentially an element in that array?

File: dvc/command/metrics.py [hunk 7]
```diff
@@ -12,7 +12,46 @@ def _show(self, metrics):
...
+            logger.info("\t{}: {}".format(fname, metric))
+            return
+
+        logger.info("\t{}:".format(fname))
+
+        if self._is_json(metric):
```
Comment (shcheklein) [inline dvc/command/metrics.py]: what was wrong with using `typ`, btw?

File: dvc/command/metrics.py [hunk 8]
```diff
@@ -12,7 +12,46 @@ def _show(self, metrics):
...
+            for line in metric.split("\n"):
```
Comment (shcheklein) [inline dvc/command/metrics.py]: why do we need this split? :) I'm not sure I understand

File: dvc/command/metrics.py [hunk 9]
```diff
@@ -12,7 +12,46 @@ def _show(self, metrics):
...
+                logger.info("\t\t{}".format(line))
+            return
+
+        reader = csv.reader(
+            metric.split("\n"), dialect=csv.Sniffer().sniff(metric)
```
Comment (shcheklein) [inline dvc/command/metrics.py]: why split("\n") here?

Comment (mroutis) [reply dvc/command/metrics.py]: The `csv.reader` can't detect newlines in the content, actually, the first argument should be an iterator, and I think file objects uses `readline()` internally as an iterator.

File: dvc/command/metrics.py (hunk 8 above)
Comment (mroutis) [reply dvc/command/metrics.py]: Because we want to print the JSON file line by line to create the following output:
```
	metrics.json:
		{
		   "deviation_mse": [
		      "0.173461",
		      "0.289545",
		      "0.297848"
		   ],
		   "data_set": [
		      "train",
		      "testing",
		      "validation"
		   ],
		   "value_mse": [
		      "0.421601",
		      "0.67528",
		      "0.671502"
		   ]
		}
```

If we don't do it, there's no way that I know to append `\t\t` to every line.

File: dvc/command/metrics.py (hunk 7 above)
Comment (mroutis) [reply dvc/command/metrics.py]: I thought that I was breaking some boundaries by taking out type from the `show` method.

What I'm planning to do, is to return the output already formatted from `repo.metrics.show`, that way, I keep respecting the interface and have access to `typ`.

Kudos to @efiop for the suggestion ^

File: dvc/command/metrics.py (hunk 6 above)
Comment (mroutis) [reply dvc/command/metrics.py]: It could be @shcheklein , but I think it is a rare case -- I thought metrics have a numeric nature, maybe I'm a little bit bias and don't taking into account everything.

File: dvc/command/metrics.py (hunk 3 above)
Comment (mroutis) [reply dvc/command/metrics.py]: @shcheklein , the current implementation with the `csv.Sniffer` has kind of a "magic setup" (it detects the properties of each line to determine the delimiter and stuff), so I think we could use CSV reader with both TSV and CSV.

File: dvc/repo/metrics/show.py [hunk 10]
```diff
@@ -66,13 +66,42 @@ def _read_typed_metric(typ, xpath, fd):
     return ret
 
 
+def _format_csv(content, delimiter):
+    import csv
+
+    reader = csv.reader(content.split("\n"), delimiter=builtin_str(delimiter))
```
Comment (shcheklein) [inline dvc/repo/metrics/show.py]: I'm not sure I understand this. Why do we need this `split('\n')` here?

File: dvc/repo/metrics/show.py [hunk 11]
```diff
@@ -66,13 +66,42 @@ def _read_typed_metric(typ, xpath, fd):
...
+            for entry, width in zip(row, max_widths)
+        )
+        for row in rows
+    ]
+
+    return "\n".join(lines)
+
+
+def _format_output(content, typ):
+    if "\n" not in content or not typ:
```
Comment (shcheklein) [inline dvc/repo/metrics/show.py]: This kind of checks are not safe for CSV that might contain entries with newlines. Could we try to get rid of the `"\n" not in content` condition and handle it with a regular `_format_csv`, even if it's a single line

File: dvc/repo/metrics/show.py [hunk 12]
```diff
@@ -86,6 +115,23 @@ def _read_metric(fd, typ=None, xpath=None, rel_path=None, branch=None):
...
+            following (raw|json|tsv|htsv|csv|hcsv).
+        xpath (str): Path to search for.
+        branch (str): Branch to look up for metrics.
+
+    Returns:
+        list(tuple): (output, typ, xpath)
+            - output:
+            - typ:
+            - xpath:
+    """
```
Comment (shcheklein) [inline dvc/repo/metrics/show.py]: ❤️

File: dvc/repo/metrics/show.py [hunk 13]
```diff
@@ -126,6 +172,21 @@ def _read_metrics_filesystem(path, typ, xpath, rel_path, branch):
...
+
+    Returns:
+        A dict mapping keys with metrics path name and content.
+        For example:
+
+        {'metric.csv': ("value_mse,deviation_mse,data_set\n"
+                        "0.421601,0.173461,train\n"
+                        "0.67528,0.289545,testing\n"
+                        "0.671502,0.297848,validation\n")}
+    """
```
Comment (shcheklein) [inline dvc/repo/metrics/show.py]: ❤️

File: dvc/repo/metrics/show.py (hunk 10 above)
Comment (mroutis) [reply dvc/repo/metrics/show.py]: @shcheklein , check this out:

```python
import csv

print("Without splitting new line")
print("--------------------------")
for line in csv.reader("x,y\n1,2"):
    print(line)

print("")
print("Splitting")
print("---------")
for line in csv.reader("x,y\n1,2".split("\n")):
    print(line)

print("")
print("File with new lines on entries")
print("------------------------------")
with open("file.csv", "w") as fobj:
    fobj.write('x,y\n"multi\nline\nstring","single-line"')

with open("file.csv", "r") as fobj:
    for line in csv.reader(fobj):
        print(line)

print("")
print("Content with new lines on entries")
print("---------------------------------")
for line in csv.reader('x,y\n"multi\nline\nstring","single line"'.split("\n")):
    print(line)
```

```
Without splitting new line
--------------------------
['x']
['', '']
['y']
[]
['1']
['', '']
['2']

Splitting
---------
['x', 'y']
['1', '2']

File with new lines on entries
------------------------------
['x', 'y']
['multi\nline\nstring', 'single-line']

Content with new lines on entries
---------------------------------
['x', 'y']
['multilinestring', 'single line']
```

This is because the CSV reader expects an iterator as the first argument.
You don't need to "split" anything when you use a file object (e.g. `with open('file.csv', 'r') as fobj: csv.reader(fobj)`), because python's file object can be used as an iterator (under the hood, the iterator is calling something like `readline()`, and it can deal with "quoted" new lines).

File: dvc/repo/metrics/show.py (hunk 11 above)
Comment (mroutis) [reply dvc/repo/metrics/show.py]: :+1:

**Comment by mroutis** [comment]
> Please, test it on CSVs that have items with newlines. It looks this is the only edge case that is left to fix.

@shcheklein , do we need to support this use case? In what case are multi-line metrics useful, and why would they choose a format like CSV instead of raw text?

File: dvc/repo/metrics/show.py [hunk 14]
```diff
@@ -66,13 +66,76 @@ def _read_typed_metric(typ, xpath, fd):
...
+        typ (str): The type of metric -- (raw|json|tsv|htsv|csv|hcsv).
+
+    Returns:
+        str: Content in a raw or tabular format.
+    """
+    if not typ:
+        return content
+    if typ == "csv" or typ == "hcsv":
+        return _format_csv(content, delimiter=",")
+    elif typ == "tsv" or typ == "tsv":
```
Comment (shcheklein) [inline dvc/repo/metrics/show.py]: `or typ == 'htsv'`

File: dvc/repo/metrics/show.py [hunk 15]
```diff
@@ -66,13 +66,76 @@ def _read_typed_metric(typ, xpath, fd):
...
+
+    Args:
+        content (str): The content of a metric.
+        typ (str): The type of metric -- (raw|json|tsv|htsv|csv|hcsv).
+
+    Returns:
+        str: Content in a raw or tabular format.
+    """
+    if not typ:
+        return content
```
Comment (shcheklein) [inline dvc/repo/metrics/show.py]: 👍

File: dvc/repo/metrics/show.py (hunk 10 above)
Comment (shcheklein) [reply dvc/repo/metrics/show.py]: Makes sense! I didn't know about `csv.reader` expecting an iterator. Consider using a safer choice to split lines - https://docs.python.org/2/library/stdtypes.html#str.splitlines

Comment (shcheklein) [reply dvc/repo/metrics/show.py]: How about this, btw:

```>>> for line in csv.reader(StringIO('"multi\nline\nstring","single line"\nx,y')):
...     print(line)
...
['multi\nline\nstring', 'single line']
['x', 'y']```?

Comment (shcheklein) [reply dvc/repo/metrics/show.py]: Using StrinIO seems a better way of doing this. Please, test all this stuff on unicode as well. Just to make sure that we don't need to provide utf-8 encoding on Python2 or something like that.

**Comment by shcheklein** [comment]
@mroutis good question! I would say yes, at leas it seems weird to remove newlines from the text completely. If there are some problems with printing it, we may consider replacing newlines with spaces when we print.

File: dvc/repo/metrics/show.py (hunk 10 above)
Comment (mroutis) [reply dvc/repo/metrics/show.py]: :+1:  Awesome, @shcheklein , will fix it on the next patch

**Comment by efiop** [comment]
@mroutis Looks like py3 builds are failing due unicode-related issues.

File: dvc/repo/metrics/show.py [hunk 16]
```diff
@@ -66,13 +66,81 @@ def _read_typed_metric(typ, xpath, fd):
...
+        "value_mse  deviation_mse   data_set\n"
+        "0.421601   0.173461        train\n"
+        "0.67528    0.289545        testing\n"
+        "0.671502   0.297848        validation\n"
+
+    """
+    import csv
+
+    delimiter = builtin_str(delimiter)
+    content = content.encode("UTF-8").splitlines(True)
```
Comment (shcheklein) [inline dvc/repo/metrics/show.py]: why not SrtringIO after all? It looks like it would handle new lines inside items very well.

Comment (mroutis) [reply dvc/repo/metrics/show.py]: @shcheklein , could you explain a little bit more how would you handle this with `StringIO`?
I haven't used that module before and from the documentation, it looks like its purpose is to handle strings in a way similar to a file object.
I'm assuming that you are thinking about replacing `content.splitlines` with a `StringIO object` to pass it to the `csv.reader`, however, StringIO doesn't call `readlines` when you use the object as an iterator.

If you use `buffer.getvalue()` it converts it to a unicode, and it doesn't have a `splitlines` method :thinking:

Comment (shcheklein) [reply dvc/repo/metrics/show.py]: Sure! A pretty reasonable explanation you can find here: https://stackoverflow.com/questions/21843693/creating-stream-to-iterate-over-from-string-in-python. It implements a line by line iterator. So it should look like a file descriptor when it's used in `csv.reader`.

Comment (shcheklein) [reply dvc/repo/metrics/show.py]: Check something like this:

```python
from dvc.utils.compat import StringIO
for line in csv.reader(StringIO('x,y\n"multi\nline\nstring","single line"')):
    print(line)
```

Comment (shcheklein) [reply dvc/repo/metrics/show.py]: If we don't want to deal with newline (that can potentially break the nice output), we should at lease replace them with a single space before we print them. You can apply `map` with a replace function probably to list of items.

**Comment by mroutis** [comment]
@shcheklein , I take into account your comments, thank you for pointing out the Unicode support. Tests are passing

**Comment by yfarjoun** [comment]
thanks, that was amazing! 🎉
